app/vmstorage/servers/vmselect.go

```diff
@@ -249,10 +257,36 @@ func getMaxMetrics(sq *storage.SearchQuery) int {
 	maxMetrics := sq.MaxMetrics
 	maxMetricsLimit := *maxUniqueTimeseries
 	if maxMetricsLimit <= 0 {
-		maxMetricsLimit = 2e9
+		maxMetricsLimit = GetMaxUniqueTimeSeries()
 	}
 	if maxMetrics <= 0 || maxMetrics > maxMetricsLimit {
 		maxMetrics = maxMetricsLimit
 	}
 	return maxMetrics
 }
+
+// GetMaxUniqueTimeSeries returns the max metrics limit calculated by available resources.
+// The calculation is split into calculateMaxUniqueTimeSeriesByResource for unit testing.
+func GetMaxUniqueTimeSeries() int {
+	once.Do(func() {
+		defaultMaxUniqueTimeseries = *maxUniqueTimeseries
+		if defaultMaxUniqueTimeseries <= 0 {
+			defaultMaxUniqueTimeseries = calculateMaxUniqueTimeSeriesByResource(*maxConcurrentRequests, memory.Remaining())
+			logger.Infof("limiting -search.maxUniqueTimeseries to %d according to -search.maxConcurrentRequests=%d and remaining memory=%d bytes. To increase the limit, reduce -search.maxConcurrentRequests or increase memory available to the process.", defaultMaxUniqueTimeseries, *maxConcurrentRequests, memory.Remaining())
+		}
+	})
+	return defaultMaxUniqueTimeseries
+}
+
+// calculateMaxUniqueTimeSeriesByResource calculate the max metrics limit calculated by available resources.
+func calculateMaxUniqueTimeSeriesByResource(maxConcurrentRequests, remainingMemory int) int {
+	if maxConcurrentRequests <= 0 {
+		// This line should NOT be reached unless the user has set an incorrect `search.maxConcurrentRequests`.
+		// In such cases, fallback to unlimited.
+		logger.Warnf("limiting -search.maxUniqueTimeseries to %v because -search.maxConcurrentRequests=%d.", 2e9, maxConcurrentRequests)
```

Comment:
Looks like a dup information to me.

Reply:
It's intended. But it absolutely makes sense to simplify it to one. https://github.com/VictoriaMetrics/VictoriaMetrics/pull/6961/commits/66cc42758ba5beabcee20b1e903d8fc6262f97b8

vmsingle PR is also updated: https://github.com/VictoriaMetrics/VictoriaMetrics/pull/7248/commits/04ea7de4f4df9638e9de35e65b11f77eb8e4e298